overlord/state/change_test.go

```diff
@@ -929,3 +929,120 @@ func (ts *taskRunnerSuite) TestAbortUnreadyLanes(c *C) {
 		c.Assert(strings.Join(obtained, " "), Equals, strings.Join(expected, " "), Commentf("setup: %s", test.setup))
 	}
 }
+
+// setup is a list of tasks "<task1> <task2>", order is <task1>-><task2>
+// (implies task2 waits for task 1)
+var cyclicDependencyTests = []struct {
+	setup string
+	order string
+	err   string
+}{
+
+	// Some basics.
+	{
+		setup: "t1",
+	}, {
+		setup: "",
+	}, {
+		// independent tasks
+		setup: "t1 t2 t3",
+	}, {
+		// simple loop
+		setup: "t1 t2",
+		order: "t1->t2 t2->t1",
+		err:   `dependency cycle involving tasks \[t1\(1\) t2\(2\)\]`,
+	},
+
+	// t1 => t2 => t3 => t4
+	// t5 => t6 => t7 => t8
+	{
+		setup: "t1 t2 t3 t4 t5 t6 t7 t8",
+		order: "t1->t2 t2->t3 t3->t4 t5->t6 t6->t7 t7->t8",
+	},
+	//               => t21 => t22
+	//             /               \
+	// t11 => t12                    => t41 => t42
+	//             \               /
+	//               => t31 => t32
+	{
+		setup: "t11 t12 t21 t22 t31 t32 t41 t42",
+		order: "t11->t12 t12->t21 t12->t31 t21->t22 t31->t32 t22->t41 t32->t41 t41->t42",
+	},
+
+	// t11 (1) => t12 (1)
+	// t21 (2) => t22 (2)
+	// t31 (3) => t32 (3)
+	// t41 (4) => t42 (4)
+	{
+		setup: "t11 t12 t21 t22 t31 t32 t41 t42",
+		order: "t11->t12 t21->t22 t31->t32 t41->t42",
+	},
+	// auto refresh like arrangement
+	//
+	//                                                  (apps)
+	//                                            => t31 (3) => t32 (3)
+	//     (snapd)               (base)         /
+	// t11 (1) => t12 (1) => t21 (2) => t22 (2)
+	//                                          \
+	//                                            => t41 (4) => t42 (4)
+	{
+		setup: "t11 t12 t21 t22 t31 t32 t41 t42",
+		order: "t11->t12 t12->t21 t21->t22 t22->t31 t22->t41 t31->t32 t41->t42",
+	},
+	// arrangement with a cyclic dependency between tasks
+	//
+	//                        /-----------------------------------------\
+	//                        |                                         |
+	//                        |                   => t31 (3) => t32 (3) /
+	//     (snapd)            v  (base)         /
+	// t11 (1) => t12 (1) => t21 (2) => t22 (2)
+	//                                          \
+	//                                            => t41 (4) => t42 (4)
+	{
+		setup: "t11 t12 t21 t22 t31 t32 t41 t42",
+		order: "t11->t12 t12->t21 t21->t22 t22->t31 t22->t41 t31->t32 t41->t42 t32->t21",
+		err:   `dependency cycle involving tasks \[t21\(3\) t22\(4\) t31\(5\) t32\(6\) t41\(7\) t42\(8\)\]`,
```

Comment:
One idea for a possible improvement in a follow-up: currently the error message includes the bottom branch because it's pointed to by a node in the cycle, even though it's not a part of the cycle itself. We could avoid this by iterating on the unsatisfiable nodes and checking if they have successors. If they don't we can remove them until nothing is removed. Then we'd be left with nodes that are strictly in the cycle.

Reply:
Great idea, thanks.